### PR TITLE
Feature/shapelets

### PR DESCRIPTION
**Pull request by Jammy2211**
Implement Shapelets in their various forms.

Example scripts here:

https://github.com/Jammy2211/autogalaxy_workspace/blob/main/scripts/imaging/modeling/advanced/light_shapelets.py

https://github.com/Jammy2211/autolens_workspace/blob/main/scripts/imaging/modeling/advanced/shapelets.py

### CODE REVIEW

File: autogalaxy/config/priors/light/shapelets/exponential.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
+ShapeletExponential:
+  centre_0:
+    type: Gaussian
+    mean: 0.0
+    sigma: 0.3
+    lower_limit: -inf
+    upper_limit: inf
```
Comment (rhayes777) [inline autogalaxy/config/priors/light/shapelets/exponential.yaml:7]: gaussian limits are implicitly infinite

File: autogalaxy/profiles/light/shapelets/exponential.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            The order of the shapelets basis function in the y-direction.
+        n_x
+            The order of the shapelets basis function in the x-direction.
+        centre
+            The (y,x) arc-second coordinates of the profile (shapelet) centre.
+        beta
+            The characteristic length scale of the shapelet basis function, defined in arc-seconds.
+        """
+
+        super().__init__(n=n, m=m, centre=centre, ell_comps=(0.0, 0.0), beta=beta)
```
Comment (rhayes777) [inline autogalaxy/profiles/light/shapelets/exponential.py:136]: Not sure if there's any point in this constructor?

File: autogalaxy/profiles/light/shapelets/polar.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+            The order of the shapelets basis function in the y-direction.
+        n_x
+            The order of the shapelets basis function in the x-direction.
+        centre
+            The (y,x) arc-second coordinates of the profile (shapelet) centre.
+        beta
+            The characteristic length scale of the shapelet basis function, defined in arc-seconds.
+        """
+
+        super().__init__(n=n, m=m, centre=centre, ell_comps=(0.0, 0.0), beta=beta)
```
Comment (rhayes777) [inline autogalaxy/profiles/light/shapelets/polar.py:134]: Same with this one. You could just put docs at the class level and not bother with the init

File: test_autogalaxy/profiles/light/shapelets/test_exponential.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+    assert image == pytest.approx(np.array([0.05784, 0.17962]), 1e-4)
+
+    shapelet = ag.lp_shapelets.ShapeletExponential(
+        n=2, m=0, centre=(0.2, 0.4), beta=1.0
+    )
+
+    image = shapelet.image_2d_from(grid=np.array([[0.0, 1.0], [0.5, 0.25]]))
+
+    assert image == pytest.approx(np.array([0.16136, 0.22476]), 1e-4)
```
Comment (rhayes777) [inline test_autogalaxy/profiles/light/shapelets/test_exponential.py:26]: Could be separated into two separate tests

File: test_autogalaxy/profiles/light/shapelets/test_polar.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
+from __future__ import division, print_function
+import numpy as np
+import pytest
+
+import autogalaxy as ag
+
+grid = np.array([[1.0, 1.0], [2.0, 2.0], [3.0, 3.0], [2.0, 4.0]])
```
Comment (rhayes777) [inline test_autogalaxy/profiles/light/shapelets/test_polar.py:7]: Slightly safer to define this as a fixture in case the array is mutated in a test

**Review by rhayes777 — APPROVED** [review]

File: autogalaxy/config/priors/light/shapelets/exponential.yaml (hunk 1 above)
Comment (Jammy2211) [reply autogalaxy/config/priors/light/shapelets/exponential.yaml:7]: Like to keep them there so users know they exist (I know this is source code but principle still applies).

**Pull request merged**